### PR TITLE
feat(embodied): Support full-shard and full-param finetune for openpi

### PR DESCRIPTION
**Pull request by LiuZhihao2022**
### Description
(1) Adapt the OpenPI model to support full-parameter fine-tuning under the FSDP framework, with compatibility for full-shard mode.
(2) Support gradient norm clipping with mixed-precision gradients. Mainly follows the official PyTorch FSDP implementation (FullyShardedDataParallel source).

### Additional information (optional, e.g., figures and logs):
Experimental evaluation on LIBERO-Goal using π₀ shows that both the training curves and training metrics are consistent with the original results.
fd87626b-87e4-4a44-98ff-9bf95fa71a32
The original no_shard mode also runs correctly.
8bec79d3-687e-4a57-aeba-ea63fbc56446

Full-parameter fine-tuning can improve the success rate.
3a12b817-c5fd-4b00-879a-bc81bece6ebf

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/models/__init__.py [hunk 1]
```diff
@@ -378,9 +378,24 @@ def get_model(cfg: DictConfig, override_config_kwargs=None):
...
+def _set_module_names(model):
+    """
+    Set the 'name' attribute for all modules in the model.
+    The name is set to the last part of the module's path (e.g., "action_in_proj").
+    This allows FSDP wrap policy to identify modules by name without using id() mapping.
+    """
+    for name, module in model.named_modules():
+        # Set name to the last part of the path (e.g., "model.action_in_proj" -> "action_in_proj")
+        path_parts = name.split(".")
+        module.name = path_parts[-1] if path_parts else name
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/models/__init__.py]: This is the expected behavior; otherwise, we would need to distinguish each replicated module.

File: rlinf/hybrid_engines/fsdp/strategy/fsdp.py [hunk 2]
```diff
@@ -228,15 +230,103 @@ def clip_grad_norm_(
 
         Args:
             - model (FSDP): The FSDP wrapped model.
-            - norm_type (Union[float,int]): The type of the used p-norm.
+            - norm_type (Union[float, int]): The type of the used p-norm.
 
         Returns:
             - float: The total norm of the gradients before clipping.
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/hybrid_engines/fsdp/strategy/fsdp.py:234]: No need

File: rlinf/models/embodiment/openpi_action_model.py [hunk 3]
```diff
@@ -75,13 +75,34 @@ class OpenPi0ForRLActionPrediction(BasePolicy, PI0Pytorch):
     @property
     def _no_split_modules(self) -> list[str]:
         # Currently, PaliGemmaForConditionalGeneration only support DDP, as many of it's modules are called without forward
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/models/embodiment/openpi_action_model.py:77]: Revised accordingly.

File: rlinf/hybrid_engines/fsdp/utils.py [hunk 4]
```diff
@@ -489,6 +503,38 @@ def get_grad_norm(
...
+    grads = [param.grad.detach().to(torch.float32) for param in params_with_grad]
+    # Compute the gradient norm in FP32, where we treat the gradients as a
+    # single vector
+    grad_norm = torch.linalg.vector_norm(
+        torch.stack(
+            [
+                torch.linalg.vector_norm(grad.detach(), norm_type, dtype=torch.float32)
+                for grad in grads
+            ],
+        ),
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/hybrid_engines/fsdp/utils.py]: It is better to follow the original PyTorch FSDP implementation.

File: rlinf/hybrid_engines/fsdp/strategy/fsdp.py [hunk 5]
```diff
@@ -228,15 +230,103 @@ def clip_grad_norm_(
...
+            total_norm = local_sharded_norm**norm_type
+            torch.distributed.all_reduce(
+                total_norm, op=torch.distributed.ReduceOp.SUM, group=self._dp_group
+            )
+            if local_nonsharded_norm is not None:
+                total_norm += local_nonsharded_norm**norm_type
+            total_norm = total_norm ** (1.0 / norm_type)
+
+        grad_norm = float(total_norm.item())
+        clip_coef = max_norm / (total_norm + 1e-6)
```
[the remark this answers is not in the document]
Comment (LiuZhihao2022) [reply rlinf/hybrid_engines/fsdp/strategy/fsdp.py]: Rectified accordingly.

**Review by guozhen1997 — APPROVED** [review]
